sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md

```diff
@@ -0,0 +1,436 @@
+# Form Recognizer Design Overview
+
+The Form Recognizer client library provides two clients to interact with the service: `FormRecognizerClient` and 
+`CustomFormClient`, which can be imported from the `azure.ai.formrecognizer` namespace. The asynchronous clients 
+can be imported from `azure.ai.formrecognizer.aio`.
+
+`FormRecognizerClient` provides methods for interacting with the prebuilt models (receipt and layout).
+`CustomFormClient` provides the methods for training custom models to analyze forms.
+
+Authentication is achieved by passing an instance of `CognitiveKeyCredential("<api_key>")` to the client,
+or by providing a token credential from `azure.identity` to use Azure Active Directory.
+
+## Prebuilt
+
+The prebuilt models are accessed through the `FormRecognizerClient`. The input form or document can be passed as a 
+string url/path to the image, or as a file stream. The SDK will determine content-type and send the appropriate header. 
+
+Both prebuilt methods return poller objects which are used to get the result.
+The `begin_extract_receipt` method returns an `ExtractedReceipt` with hardcoded receipt fields.
+The `begin_extract_layout` method returns the extracted tables in a tabular format such that the user can
+index into a specific row or column and easily integrate with other Python libraries.
+
+If the keyword argument `include_text_details=True` is passed in, the `raw_` attributes will be populated with the
+raw OCR result for each value/cell.
+
+### Form Recognizer Client
+```python
+from azure.ai.formrecognizer import FormRecognizerClient
+
+client = FormRecognizerClient(endpoint: str, credential: Union[CognitiveKeyCredential, TokenCredential])
+
+client.begin_extract_receipt(form: Any, **kwargs) -> LROPoller -> List[ExtractedReceipt]
+
+client.begin_extract_layout(form: Any, **kwargs) -> LROPoller -> List[ExtractedLayoutPage]
+```
+
+### Receipt Models
+```python
+class ExtractedReceipt:
+    receipt_items: List[ReceiptItem]
+    merchant_address: str
+    merchant_name: str
+    merchant_phone_number: str
+    receipt_type: str
+    subtotal: float
+    tax: float
+    tip: float
+    total: float
+    transaction_date: str
+    transaction_time: str
+    fields: ReceiptFields
```

Comment:
These fields will come back in the deserialized response via the gen code. Is this a perf concern? Or are 
you saying to hide these fields to not overwhelm the user?

Reply:
Yes, mainly perf, since the information will be duplicated (or triplicated since this is a class object and the original response is presumably sitting in a dict).

What do you think about using generics here?

Reply:
We will not duplicate the values. The only (memory) overhead would be bookkeeping data. We'll measure this and make sure that we don't have issues/understand the overhead. 

What are you referring to with "generics"?

Reply:
Sure, for generics:

`a`, `b`, and `c` are statically typed to `int`, `str`, and `datetime`, respectively. It's still dynamic typing, but the static hints allow a sensible IDE or analysis tool to understand the value type.

```python
from datetime import datetime
from typing import Generic, TypeVar

T = TypeVar('T')


class FieldValue(Generic[T]):
    def __init__(self, value: T):
        self.value = value


if __name__ == '__main__':
    field1 = FieldValue(value=123)
    field2 = FieldValue(value='hello')
    field3 = FieldValue(value=datetime.utcnow())

    a = field1.value
    b = field2.value
    c = field3.value
```

This was written in 3.7, but I think it should work as far back as 3.5 too.